ambry-server/src/main/java/com/github/ambry/server/RepairRequestsDB.java

```diff
@@ -0,0 +1,268 @@
+      logger.warn("Unable to connect to endpoint {} due to {}", dbEndpoint.getUrl(), e.getMessage());
+      throw e;
+    }
+  }
+
+  /**
+   * Remove one {@link RepairRequestRecord} from the database
+   * @param blobId the blob id
+   * @param operationType the operation time, either TtlUpdate or Delete
+   */
+  public void removeRepairRequests(String blobId, short operationType) throws SQLException {
+    // private static final String DELETE_QUERY = String.format(""
+    //     + "DELETE FROM %s "
+    //     + "WHERE %s=? AND %s=?",
+    //     REPAIR_REQUESTS_TABLE,
+    //     BLOB_ID, OPERATION_TYPE);
+    try {
+      PreparedStatement statement = connection.prepareStatement(DELETE_QUERY);
+      statement.setBytes(1, Base64.decodeBase64(blobId));
+      statement.setShort(2, operationType);
+      statement.executeUpdate();
+    } catch (SQLException e) {
+      logger.warn("failed to delete record from {} due to {}", dbEndpoint.getUrl(), e.getMessage());
+      throw e;
+    }
+  }
+
+  /**
+   * Insert one {@link RepairRequestRecord}
+   * @param record the record to insert
+   */
+  public void putRepairRequests(RepairRequestRecord record) throws SQLException {
+    // private static final String INSERT_QUERY = String.format(""
+    //    + "INSERT INTO %1$s "
+    //    + "(%2$s, %3$s, %4$s, %5$s, %6$s, %7$s, %8$s, %9$s) VALUES (?, ?, ?, ?, ?, ?, ?, ?)",
+    //    REPAIR_REQUESTS_TABLE,
+    //    BLOB_ID, PARTITION_ID, SOURCE_HOST_NAME, SOURCE_HOST_PORT, OPERATION_TYPE, OPERATION_TIME, LIFE_VERSION, EXPIRATION_TYPE);
+    try {
+      PreparedStatement statement = connection.prepareStatement(INSERT_QUERY);
+      statement.setBytes(1, Base64.decodeBase64(record.getBlobId()));
+      statement.setInt(2, record.getPartitionId());
+      statement.setString(3, record.getSourceHostName());
+      statement.setInt(4, record.getSourceHostPort());
+      statement.setShort(5, record.getOperationType());
+      statement.setTimestamp(6, new Timestamp(record.getOperationTimeMs()));
+      statement.setShort(7, record.getLifeVersion());
+      if (record.getExpirationTimeMs() != Utils.Infinite_Time) {
+        statement.setTimestamp(8, new Timestamp(record.getExpirationTimeMs()));
+      } else {
+        statement.setTimestamp(8, null);
+      }
+      statement.executeUpdate();
+    } catch (SQLException e) {
+      logger.warn("failed to insert record to {} due to {}", dbEndpoint.getUrl(), e.getMessage());
+      throw e;
+    }
+  }
+
+  /**
+   * Select the records from one partition ordered by the operation time.
+   * @param partitionId partition id
+   * @return the oldest {@link RepairRequestRecord}s.
+   */
+  public List<RepairRequestRecord> getRepairRequests(int partitionId) throws SQLException {
+    // private static final String GET_QUERY = String.format(""
+    //    + "SELECT %s, %s, %s, %s, %s, %s, %s, %s "
+    //    + "FROM %s "
+    //    + "WHERE %s = ? "
+    //    + "ORDER BY %s ASC "
```

Comment:
QQ: if for some reason some request can't be repaired, will that be stay in the REPAIR_REQUESTS_TABLE forever? Should we automatically purge the data if creationTime < currentTime - retentionTime?

Reply:
For any reason if the request cannot be repaired, it's critical. We need retry until it's fixed or get oncall's attention for it.
But you are right. we may need continue to fix other repair requests if possible, I may add continue search by page function. Right now I added one TODO. I'll add it in the following PR.
// LOCAL_CONSISTENCY_TODO continue the Get from one token